### PR TITLE
migrate beefy relayer into own worker

### PR DESCRIPTION
**Pull request by musnit**
all seems to be working, with initial and complete transactions all being sent :)

e2e tests all pass

it's based off of the parachainCommitmentRelayer branch for convenience and I expect that will be merged first

@denalimarsh plz review thoroughly

### CODE REVIEW

File: relayer/workers/beefyrelayer/main.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	beefyMessages           chan store.BeefyRelayInfo
+	ethHeaders              chan chain.Header
+}
+
+const Name = "beefy-relayer"
+
+func NewWorker(relaychainConfig *relaychain.Config, ethereumConfig *ethereum.Config, dbConfig *store.Config) (*Worker, error) {
+	log := logrus.WithField("beefy-relayer", Name)
+
+	fmt.Println("Creating beefy-relayer")
```
Comment (denalimarsh) [inline relayer/workers/beefyrelayer/main.go:38]:
```suggestion
	log.Info("Creating beefy-relayer")
```

File: relayer/workers/beefyrelayer/main.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+		beefyRelaychainListener: beefyRelaychainListener,
+		log:                     log,
+		beefyDB:                 beefyDB,
+		beefyMessages:           beefyMessages,
+		ethHeaders:              ethHeaders,
+	}, nil
+}
+
+func (worker *Worker) Start(ctx context.Context, eg *errgroup.Group) error {
+	fmt.Println("Starting beefy-relayer")
```
Comment (denalimarsh) [inline relayer/workers/beefyrelayer/main.go:96]:
```suggestion
	worker.log.Info("Starting beefy-relayer")
```

File: relayer/workers/beefyrelayer/beefy-ethereum-writer.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+	beefyDB           *store.Database
+	lightClientBridge *lightclientbridge.Contract
+	databaseMessages  chan<- store.DatabaseCmd
+	beefyMessages     <-chan store.BeefyRelayInfo
+	log               *logrus.Entry
+}
+
+func NewBeefyEthereumWriter(ethereumConfig *ethereum.Config, ethereumConn *ethereum.Connection, beefyDB *store.Database,
+	databaseMessages chan<- store.DatabaseCmd, beefyMessages <-chan store.BeefyRelayInfo,
+	log *logrus.Entry) (*BeefyEthereumWriter, error) {
```
Comment (denalimarsh) [inline relayer/workers/beefyrelayer/beefy-ethereum-writer.go:31]:
```suggestion
	log *logrus.Entry) *BeefyEthereumWriter {
```
Error is always nil, are we able to remove `error` from the return values?

File: relayer/workers/beefyrelayer/beefy-ethereum-writer.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+	return &BeefyEthereumWriter{
+		ethereumConfig:   ethereumConfig,
+		ethereumConn:     ethereumConn,
+		beefyDB:          beefyDB,
+		databaseMessages: databaseMessages,
+		beefyMessages:    beefyMessages,
+		log:              log,
+	}, nil
```
Comment (denalimarsh) [inline relayer/workers/beefyrelayer/beefy-ethereum-writer.go:39]:
```suggestion
	}
```

File: relayer/workers/parachaincommitmentrelayer/parachain-commitment-listener.go [hunk 5]
```diff
@@ -15,22 +15,25 @@ import (
...
 )
 
 type ParachainCommitmentListener struct {
-	config   *parachain.Config
-	conn     *parachain.Connection
-	messages chan<- []chain.Message
-	log      *logrus.Entry
+	config               *parachain.Config
+	parachainConnection  *parachain.Connection
+	relaychainConnection *relaychain.Connection
```
Comment (denalimarsh) [inline relayer/workers/parachaincommitmentrelayer/parachain-commitment-listener.go:25]: `relaychainConnection` is unused by this worker. Why was it added? Is it just scaffolding for future work?

File: relayer/workers/parachaincommitmentrelayer/main.go [hunk 6]
```diff
@@ -10,15 +10,18 @@ import (
...
 	"github.com/snowfork/polkadot-ethereum/relayer/crypto/secp256k1"
 	"github.com/snowfork/polkadot-ethereum/relayer/substrate"
 )
 
 type Worker struct {
 	parachainConfig             *parachain.Config
+	relaychainConfig            *relaychain.Config
 	ethereumConfig              *ethereum.Config
 	parachainConn               *parachain.Connection
+	relaychainConn              *relaychain.Connection
```
Comment (denalimarsh) [inline relayer/workers/parachaincommitmentrelayer/main.go:24]: Should `relaychainConn`/`relaychainConfig` be added to this worker in a separate PR?

File: relayer/workers/beefyrelayer/beefy-ethereum-listener.go [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+	beefyMessages     chan<- store.BeefyRelayInfo
+	dbMessages        chan<- store.DatabaseCmd
+	headers           chan<- chain.Header
+	blockWaitPeriod   uint64
+	log               *logrus.Entry
+}
+
+func NewBeefyEthereumListener(ethereumConfig *ethereum.Config, ethereumConn *ethereum.Connection, beefyDB *store.Database,
+	beefyMessages chan<- store.BeefyRelayInfo, dbMessages chan<- store.DatabaseCmd, headers chan<- chain.Header,
+	log *logrus.Entry) (*BeefyEthereumListener, error) {
```
Comment (denalimarsh) [inline relayer/workers/beefyrelayer/beefy-ethereum-listener.go:38]:
```suggestion
	log *logrus.Entry) *BeefyEthereumListener {
```

File: relayer/workers/beefyrelayer/beefy-ethereum-listener.go [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+	return &BeefyEthereumListener{
+		ethereumConfig:  ethereumConfig,
+		ethereumConn:    ethereumConn,
+		beefyDB:         beefyDB,
+		dbMessages:      dbMessages,
+		beefyMessages:   beefyMessages,
+		headers:         headers,
+		blockWaitPeriod: 0,
+		log:             log,
+	}, nil
```
Comment (denalimarsh) [inline relayer/workers/beefyrelayer/beefy-ethereum-listener.go:48]:
```suggestion
	}
```

**Review by denalimarsh — COMMENTED** [review]
LGTM, just a couple of small changes and questions.

File: relayer/workers/parachaincommitmentrelayer/parachain-commitment-listener.go (hunk 5 above)
Comment (musnit) [reply relayer/workers/parachaincommitmentrelayer/parachain-commitment-listener.go:25]: yep, sorry should have done in different PR, gonna leave it in for now as i'll be working on follow up PR today

File: relayer/workers/parachaincommitmentrelayer/main.go (hunk 6 above)
Comment (musnit) [reply relayer/workers/parachaincommitmentrelayer/main.go:24]: yep, sorry should have done in different PR, gonna leave it in for now as i'll be working on follow up PR today

**Comment by musnit** [comment]
@denalimarsh all fixed in last commit

**Review by vgeddes — APPROVED** [review]
Looks good :+1:

File: relayer/chain/ethereum/listener.go [hunk 9]
```diff
@@ -32,30 +28,23 @@ type Listener struct {
...
-	beefyMessages               chan<- store.BeefyRelayInfo
-	dbMessages                  chan<- store.DatabaseCmd
 	headers                     chan<- chain.Header
 	blockWaitPeriod             uint64
 	log                         *logrus.Entry
 }
 
-func NewListener(config *Config, conn *Connection, db *store.Database, messages chan<- []chain.Message,
-	beefyMessages chan<- store.BeefyRelayInfo, dbMessages chan<- store.DatabaseCmd, headers chan<- chain.Header,
+func NewListener(config *Config, conn *Connection, messages chan<- []chain.Message,
```
Comment (vgeddes) [inline relayer/chain/ethereum/listener.go:38]: small nit: for these long method signatures, lets break the signature along multiple lines

File: relayer/chain/relaychain/connection.go [hunk 10]
```diff
@@ -9,27 +9,32 @@ import (
...
-func NewConnection(endpoint string, kp *signature.KeyringPair, log *logrus.Entry) *Connection {
+func NewConnection(endpoint string, log *logrus.Entry) *Connection {
 	return &Connection{
 		endpoint: endpoint,
-		kp:       kp,
 		log:      log,
 	}
 }
 
+func (co *Connection) GetAPI() *gsrpc.SubstrateAPI {
```
Comment (vgeddes) [inline relayer/chain/relaychain/connection.go:30]: nit: In golang, the convention is not to prepend `Get` to getters.

File: relayer/chain/relaychain/connection.go [hunk 11]
```diff
@@ -9,27 +9,32 @@ import (
...
+	return co.api
+}
+
+func (co *Connection) GetMetadata() *types.Metadata {
```
Comment (vgeddes) [inline relayer/chain/relaychain/connection.go:34]: ditto. Just `Metadata()`

File: relayer/core/relay.go [hunk 12]
```diff
@@ -122,21 +107,26 @@ func NewRelay() (*Relay, error) {
...
-		err = ethChain.SetReceiver(subMessages, nil, dbMessages, beefyMessages)
+		err = ethChain.SetReceiver(subMessages, nil)
 		if err != nil {
 			return nil, err
 		}
 
-		err = relayChain.SetSender(subMessages, nil, beefyMessages)
+	}
+
+	parachainCommitmentRelayer := &parachaincommitmentrelayer.Worker{}
```
Comment (vgeddes) [inline relayer/core/relay.go:117]: nit: My poor eyes. Prefer short variable names such as `relayer`.

File: relayer/workers/beefyrelayer/beefy-ethereum-writer.go [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+	ethereumConfig    *ethereum.Config
+	ethereumConn      *ethereum.Connection
+	beefyDB           *store.Database
+	lightClientBridge *lightclientbridge.Contract
+	databaseMessages  chan<- store.DatabaseCmd
+	beefyMessages     <-chan store.BeefyRelayInfo
+	log               *logrus.Entry
+}
+
+func NewBeefyEthereumWriter(ethereumConfig *ethereum.Config, ethereumConn *ethereum.Connection, beefyDB *store.Database,
```
Comment (vgeddes) [inline relayer/workers/beefyrelayer/beefy-ethereum-writer.go:29]: nit: Lets break this method signature along multiple lines for easier readability.

File: relayer/workers/beefyrelayer/beefy-relaychain-listener.go [hunk 14]
```diff
@@ -12,27 +9,30 @@ import (
...
 	"github.com/snowfork/polkadot-ethereum/relayer/substrate"
+	"github.com/snowfork/polkadot-ethereum/relayer/workers/beefyrelayer/store"
 )
 
-type Listener struct {
-	config        *Config
-	conn          *Connection
-	beefyMessages chan<- store.BeefyRelayInfo
-	log           *logrus.Entry
+type BeefyRelaychainListener struct {
```
Comment (vgeddes) [inline relayer/workers/beefyrelayer/beefy-relaychain-listener.go:17]: nit: If this struct is already in a `beefyrelayer` package, can just shorten the struct name to `Listener`. Is easier to read.

File: relayer/workers/beefyrelayer/beefy-relaychain-listener.go [hunk 15]
```diff
@@ -12,27 +9,30 @@ import (
...
+	return &BeefyRelaychainListener{
+		relaychainConfig: relaychainConfig,
+		relaychainConn:   relaychainConn,
+		beefyMessages:    beefyMessages,
+		log:              log,
 	}
 }
 
-func (li *Listener) Start(ctx context.Context, eg *errgroup.Group) error {
+func (li *BeefyRelaychainListener) Start(ctx context.Context, eg *errgroup.Group) error {
```
Comment (vgeddes) [inline relayer/workers/beefyrelayer/beefy-relaychain-listener.go:35]: ditto: No need to rename this to `BeefyRelaychainListener`.

File: relayer/workers/beefyrelayer/main.go [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+	"github.com/snowfork/polkadot-ethereum/relayer/chain/relaychain"
+	"github.com/snowfork/polkadot-ethereum/relayer/crypto/secp256k1"
+	"github.com/snowfork/polkadot-ethereum/relayer/workers/beefyrelayer/store"
+)
+
+type Worker struct {
+	relaychainConfig        *relaychain.Config
+	ethereumConfig          *ethereum.Config
+	relaychainConn          *relaychain.Connection
+	beefyEthereumListener   *BeefyEthereumListener
```
Comment (vgeddes) [inline relayer/workers/beefyrelayer/main.go:23]: nit: Don't need to prepend `beefy` to some of these field names.

**Review by vgeddes — APPROVED** [review]
Added some small nits but otherwise looks good 🚀

File: relayer/core/relay.go (hunk 12 above)
Comment (musnit) [reply relayer/core/relay.go:117]: yeh, now that things are in their own workers, everything is contextual and so we can use short names everywhere

we can do a follow up code quality/code standards refactor PR after all this stuff is merged i think

**Comment by musnit** [comment]
@vgeddes will do fixes in follow up PR - issue here: https://github.com/Snowfork/polkadot-ethereum/issues/359

File: relayer/chain/ethereum/listener.go (hunk 9 above)
Comment (musnit) [reply relayer/chain/ethereum/listener.go:38]: @vgeddes can't gofmt do this for us?

File: relayer/workers/beefyrelayer/beefy-relaychain-listener.go (hunk 14 above)
Comment (musnit) [reply relayer/workers/beefyrelayer/beefy-relaychain-listener.go:17]: i think in this context, the name is still valuable as there are multiple listeners with different purposes/sources

File: relayer/workers/beefyrelayer/beefy-relaychain-listener.go (hunk 15 above)
Comment (musnit) [reply relayer/workers/beefyrelayer/beefy-relaychain-listener.go:35]: i think in this context, the name is still valuable as there are multiple listeners with different purposes/sources
